composer/trainer/trainer.py

```diff
@@ -1176,6 +1179,11 @@ def __init__(
         # Run Event.INIT
         self.engine.run_event(Event.INIT)
 
+        # Log hparams.
+        if self.auto_log_hparams:
+            self.local_hparams = extract_hparams(locals())
```

Comment:
I think we need to deepcopy locals here, or be very careful not to modify it, otherwise could have unexpected side effects. And we probably don't want to deepcopy it, as that could be a costly operation. Thoughts?

Reply:
I think if we only log the original object if it is int, str, or float and otherwise log `obj.__class__.__name__`, then we are good because:

a. we don't log locals() directly, we make a new dictionary.
b. any obj we log directly from locals() is immutable (int, float, str).

Reply:
The first thing that happens inside of `extract_params` is that `self` is popped from the input dict though. Maybe everything else is fine if that line is changed to not pop, but rather just skip that key.

Reply:
ooohhhh good call. fixed